### PR TITLE
examples/gnrc_border_route: simplify ZEP setup on native

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

This PR makes it easier to set up a border router on `native` that will route to other `native` instances using the `socket_zep` 802.15.4 emulation.

`socket_zep` can only connect to one other node / client, so a simple dispatcher was added that broadcasts a packet to all other connected `socket_zep` clients.
For convenience, this is started automatically by the border router Makefile.

### Testing procedure

- start the native border router:

      make -C examples/gnrc_border_router BOARD=native all term

You should have two interfaces and a prefix configured via `uhcp`

```
Iface  7  HWaddr: 45:5B  Channel: 26  NID: 0x23 
          Long HWaddr: 00:5A:45:50:0A:00:45:5B 
          L2-PDU:102  MTU:1280  HL:64  RTR  
          RTR_ADV  6LO  IPHC  
          Source address length: 8
          Link type: wireless
          inet6 addr: fe80::25a:4550:a00:455b  scope: link  VAL
          inet6 addr: 2001:db8::25a:4550:a00:455b  scope: global  VAL
          inet6 group: ff02::2
          inet6 group: ff02::1
          inet6 group: ff02::1:ff00:455b
          
Iface  6  HWaddr: 6A:5A:B1:42:3A:F1 
          L2-PDU:1500  MTU:1500  HL:64  RTR  
          Source address length: 6
          Link type: wired
          inet6 addr: fe80::685a:b1ff:fe42:3af1  scope: link  VAL
          inet6 addr: fe80::2  scope: link  VAL
          inet6 group: ff02::2
          inet6 group: ff02::1
          inet6 group: ff02::1:ff42:3af1
          inet6 group: ff02::1:ff00:2
```

- start the native `soccket_zep` node

      USEMODULE=socket_zep make -C examples/gnrc_networking BOARD=native all term

You should have received a prefix from the border router 

```
Iface  7  HWaddr: 45:5A  Channel: 26  NID: 0x23 
          Long HWaddr: 00:5A:45:50:0A:00:45:5A 
          L2-PDU:102  MTU:1280  HL:64  RTR  
          RTR_ADV  6LO  IPHC  
          Source address length: 8
          Link type: wireless
          inet6 addr: fe80::25a:4550:a00:455a  scope: link  VAL
          inet6 addr: 2001:db8::25a:4550:a00:455a  scope: global  VAL
          inet6 group: ff02::2
          inet6 group: ff02::1
          inet6 group: ff02::1:ff00:455a
          
          Statistics for Layer 2
            RX packets 4  bytes 286
            TX packets 3 (Multicast: 1)  bytes 171
            TX succeeded 6 errors 0
          Statistics for IPv6
            RX packets 3  bytes 304
            TX packets 3 (Multicast: 1)  bytes 224
            TX succeeded 3 errors 0

> ping6 ff02::1
ping6 ff02::1
12 bytes from fe80::25a:4550:a00:455b%7: icmp_seq=0 ttl=64 rssi=255 dBm time=3.082 ms
12 bytes from fe80::25a:4550:a00:455b%7: icmp_seq=1 ttl=64 rssi=255 dBm time=1.833 ms
12 bytes from fe80::25a:4550:a00:455b%7: icmp_seq=2 ttl=64 rssi=255 dBm time=2.419 ms

--- ff02::1 PING statistics ---
3 packets transmitted, 3 packets received, 0% packet loss
round-trip min/avg/max = 1.833/2.444/3.082 ms
```

### CODE REVIEW

File: examples/gnrc_border_router/Makefile.native.conf [hunk 1]
```diff
@@ -1,16 +1,24 @@
...
 
 # Set CFLAGS if not being set via Kconfig
 CFLAGS += $(if $(CONFIG_KCONFIG_MODULE_DHCPV6),,-DCONFIG_DHCPV6_CLIENT_PFX_LEASE_MAX=$(ZEP_DEVICES))
 
-# -z [::1]:$PORT for each ZEP device
-TERMFLAGS += $(patsubst %,-z [::1]:%, $(shell seq $(ZEP_PORT_BASE) $(ZEP_PORT_MAX)))
+# -z [::]:$LOCAL_PORT,[::1]:$REMOTE_PORT for each ZEP device
+TERMFLAGS += $(shell paste -d ',' <(seq -f "-z [::1]:%g" $(ZEP_PORT_BASE) 2 $(ZEP_PORT_MAX))                     \
+                                  <(seq -f "[::1]:%g" $$(($(ZEP_PORT_BASE)+1)) 2 $$(($(ZEP_PORT_MAX)+1)))       \
+                                  | tr '\n' ' ')
```
Comment (miri64) [inline examples/gnrc_border_router/Makefile.native.conf:19]: PR says this simplifies things, but can you please include an explainer here what is  happening here?

File: examples/gnrc_networking/Makefile [hunk 2]
```diff
@@ -36,6 +36,11 @@ USEMODULE += netstats_rpl
 # development process:
 DEVELHELP ?= 1
 
+# set the ZEP port for native (reverse of the default border router configuration)
+ifneq (,$(filter socket_zep,$(USEMODULE)))
+  PORT ?= -z [::1]:17755,[::1]:17754
```
Comment (miri64) [inline examples/gnrc_networking/Makefile:41]: IMHO this shouldn't be `PORT`. It's already pretty murky that the TAP interface with `native` is the `PORT` (but only if there is `netdev_tap`), so adding more blurriness to that variable doesn't make things better (not to talk about the confusing fact that this PORT contains network ports...). Why not just

```suggestion
  TERMFLAGS += -z [::1]:17755,[::1]:17754
```

File: examples/gnrc_border_router/Makefile.native.conf (hunk 1 above)
Comment (benpicco) [reply examples/gnrc_border_router/Makefile.native.conf:20]: Yea I noticed I need to use different local ports if there are multiple ZEP devices.

Unfortunately the documentation for ZEP is very sparse - is extending the local port to accept multiple connections the right way forward or am I missing something here?

Comment (miri64) [reply examples/gnrc_border_router/Makefile.native.conf:20]: ZEP is UDP, there are no connections, so why should it need to be extended to accept multiple connection?

File: examples/gnrc_border_router/Makefile.native.conf [hunk 3]
```diff
@@ -1,16 +1,25 @@
...
 CFLAGS += -DSOCKET_ZEP_MAX=$(ZEP_DEVICES)
 CFLAGS += -DASYNC_READ_NUMOF=$(shell expr $(ZEP_DEVICES) + 1)
 
 # Set CFLAGS if not being set via Kconfig
 CFLAGS += $(if $(CONFIG_KCONFIG_MODULE_DHCPV6),,-DCONFIG_DHCPV6_CLIENT_PFX_LEASE_MAX=$(ZEP_DEVICES))
 
-# -z [::1]:$PORT for each ZEP device
-TERMFLAGS += $(patsubst %,-z [::1]:%, $(shell seq $(ZEP_PORT_BASE) $(ZEP_PORT_MAX)))
+# generate -z [::1]:$LOCAL_PORT,[::1]:$REMOTE_PORT argument for each ZEP device, starting with $ZEP_PORT_BASE
+# $REMOTE_PORT will be $LOCAL_PORT + 1
```
Comment (miri64) [inline examples/gnrc_border_router/Makefile.native.conf:17]: Does this really make the usage of ZEP easier? As I read the comment it is now impossible to create a pair of nodes that are reachable to each other, which I'd say is the main use case.

File: examples/gnrc_border_router/Makefile.native.conf (hunk 1 above)
Comment (benpicco) [reply examples/gnrc_border_router/Makefile.native.conf:20]: > ZEP is UDP, there are no connections, so why should it need to be extended to accept multiple connection?

￼That's what I thought first too.
But it uses `real_connect` to bind to the remote socket and then only ever will read/write from that.

Starting a second client will only yield a "connection refused".
We would also need to keep a list of a clients  to send broadcast messages, this is not done.

So I think this was originally used in a different fashion than I anticipated.

Comment (miri64) [reply examples/gnrc_border_router/Makefile.native.conf:20]: Mhh ah, because it was originally intended to be used with the server... It is indeed weird, that you _have_ to set the remote, but the local is optional.

https://github.com/RIOT-OS/RIOT/blob/f64511ddb5e6301beac5e16aae91f6fe34635f7c/cpu/native/startup.c#L298-L303

Even though I wrote that, in retrospect I find it weird.

Comment (miri64) [reply examples/gnrc_border_router/Makefile.native.conf:20]: Ah wait. If you don't have a remote, the "radio" would have no way to know where to send the packet, so the `real_connect()` allows for that and you _need_ a server to allow for >2 nodes communicating that dispatches which nodes to send to.

File: examples/gnrc_border_router/Makefile.native.conf (hunk 3 above)
Comment (benpicco) [reply examples/gnrc_border_router/Makefile.native.conf:17]: I'll change the PR and drop 40d4a6fd4e9a1f9802c04a778c0a0c2d8a685e67, modify this to use a UDP dispatcher instead.

This allows to connect multiple clients and is easier to implement than putting the logic into native / RIOT.

I'll squash since this basically re-writes the PR.

**Comment by benpicco** [comment]
I re-wrote the PR to use a dispatcher instead of setting up direct connections.
This way many nodes can connect to the border router.

File: Makefile.include [hunk 4]
```diff
@@ -726,6 +726,7 @@ termdeps: $(TERMDEPS)
 term: $(TERMDEPS)
 	$(call check_cmd,$(TERMPROG),Terminal program)
 	$(TERMPROG) $(TERMFLAGS) $(TERMTEE)
+	$(POSTTERM)
```
Comment (miri64) [inline Makefile.include:729]: Question to the testing people @fjmolinas @MrKevinWeiss, @leandrolanzieri ...: is this fine?

File: dist/tools/zep_dispatch/main.c [hunk 5]
```diff
@@ -0,0 +1,117 @@
+/*
```
Comment (miri64) [inline dist/tools/zep_dispatch/main.c:1]: Maybe unpopular proposal, but why not implement this in `python`? You wouldn't need to compile it and overall I think this would be way simpler.

**Review by miri64 — DISMISSED** [review]

File: dist/tools/zep_dispatch/main.c (hunk 5 above)
Comment (benpicco) [reply dist/tools/zep_dispatch/main.c:1]: Mostly because I'm more comfortable with C than with Python ;-)

File: examples/gnrc_border_router/Makefile.native.conf [hunk 6]
```diff
@@ -12,8 +14,10 @@ CFLAGS += $(if $(CONFIG_KCONFIG_MODULE_DHCPV6),,-DCONFIG_DHCPV6_CLIENT_PFX_LEASE
 # -z [::1]:$PORT for each ZEP device
 TERMFLAGS += $(patsubst %,-z [::1]:%, $(shell seq $(ZEP_PORT_BASE) $(ZEP_PORT_MAX)))
 
+POSTTERM ?= killall uhcpd ; killall zep_dispatch
```
Comment (fjmolinas) [inline examples/gnrc_border_router/Makefile.native.conf:17]: This is a bit aggressive, why did you need to add `uhcpd` here?

Comment (benpicco) [reply examples/gnrc_border_router/Makefile.native.conf:17]: Hm I guess alternatively I could create a `start_network.sh` script for `native`

Comment (benpicco) [reply examples/gnrc_border_router/Makefile.native.conf]: I created a script that takes care of starting / shutting down all the deamons.
Should be much cleaner now :)

File: dist/tools/zep_dispatch/start_network.sh [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+ZEP_DISPATCH_DIR="$(cd "$(dirname "$0")" && pwd -P)"
+UHCPD="$(cd "${ZEP_DISPATCH_DIR}/../uhcpd/bin" && pwd -P)/uhcpd"
+DHCPD="$(cd "${ZEP_DISPATCH_DIR}/../dhcpv6-pd_ia/" && pwd -P)/dhcpv6-pd_ia.py"
+ZEP_DISPATCH="${ZEP_DISPATCH_DIR}/bin/zep_dispatch"
+TAPSETUP="$(cd "${ZEP_DISPATCH_DIR}/../tapsetup/" && pwd -P)/tapsetup"
+
+TAP_GLB="fdea:dbee:f::1/64"
+
+NOSUDO="sudo -u ${SUDO_USER}"
```
Comment (miri64) [inline dist/tools/zep_dispatch/start_network.sh:11]: I rather would prefer a warning here. Calling `sudo` in a script, especially in such a hidden manner, seems not very nice.

Comment (miri64) [reply dist/tools/zep_dispatch/start_network.sh:11]: NVM, the `-u` actually "unsudos" the binay called.

File: dist/tools/zep_dispatch/start_network.sh [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+if [ "$1" = "-d" ] || [ "$1" = "--use-dhcpv6" ]; then
+    USE_DHCPV6=1
+    shift 1
+else
+    USE_DHCPV6=0
+fi
+
+if [ "$1" = "-z" ] || [ "$1" = "--use-zep" ]; then
```
Comment (benpicco) [inline dist/tools/zep_dispatch/start_network.sh:65]: should rather be `--use-dispatch`

File: examples/gnrc_networking/Makefile [hunk 9]
```diff
@@ -36,6 +36,11 @@ USEMODULE += netstats_rpl
 # development process:
 DEVELHELP ?= 1
 
+# set the ZEP port for native (reverse of the default border router configuration)
```
Comment (benpicco) [inline examples/gnrc_networking/Makefile:39]:
```suggestion
# set the ZEP port for native
```

File: dist/tools/zep_dispatch/start_network.sh [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+    UHCPD_PID=$!
+}
+
+start_dhcpd() {
+    DHCPD_PIDFILE=$(mktemp)
+    ${DHCPD} -d -p ${DHCPD_PIDFILE} ${TAP} ${PREFIX} 2> /dev/null
+}
+
+start_zep_dispatch() {
+    ${ZEP_DISPATCH} ::1 ${ZEP_PORT_BASE} > /dev/null &
```
Comment (miri64) [inline dist/tools/zep_dispatch/start_network.sh:54]:
```suggestion
    ${ZEP_DISPATCH} :: ${ZEP_PORT_BASE} > /dev/null &
```

?

**Review by miri64 — COMMENTED** [review]
Works (kind of) as described _after_ the border router sent something over the ZEP interface. Before it is not registered with the dispatcher, but that is to be expected.

**Review by miri64 — COMMENTED** [review]
Can you add a comment to the README maybe, that the border router needs to send something after start with `native` so it works (for now)?

File: dist/tools/zep_dispatch/start_network.sh (hunk 10 above)
Comment (benpicco) [reply dist/tools/zep_dispatch/start_network.sh]: I felt it would be 'safer' if this only listened on localhost.
But also fine with listening on all addresses.

File: dist/tools/zep_dispatch/main.c [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+            memcpy(&client->addr, &src_addr, addr_len);
+            list_add(&head, &client->node);
+        }
+    }
+}
+
+int main(int argc, char **argv)
+{
+    if (argc < 3) {
+        fprintf(stderr, "usage: %s <interface> <port>\n",
```
Comment (miri64) [inline dist/tools/zep_dispatch/main.c:82]: `<address>`

**Comment by benpicco** [comment]
8c4293383ac5e02648ddaa1075923e9e4b1f265c should provide a better solution

**Comment by miri64** [comment]
> [8c42933](https://github.com/RIOT-OS/RIOT/commit/8c4293383ac5e02648ddaa1075923e9e4b1f265c) should provide a more convenient solution.

Should we maybe move that rather to a follower-up.

File: dist/tools/zep_dispatch/start_network.sh [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+    start_zep_dispatch
+fi
+
+if [ ${USE_DHCPV6} -eq 1 ]; then
+    start_dhcpd
+else
+    start_uhcpd
+fi
+
+${NOSUDO} ${ELFFILE} ${TAP} $*
```
Comment (miri64) [inline dist/tools/zep_dispatch/start_network.sh:100]: I think this is in a race with `zep_dispatch`... Since you only provide the remote pair, `socket_zep` will set its local port to 17754 (= `ZEP_PORT_BASE`):

https://github.com/RIOT-OS/RIOT/blob/76ddfdbbdef0ad0df8a2528381046233a2b6e21a/cpu/native/include/socket_zep_params.h#L41

https://github.com/RIOT-OS/RIOT/blob/76ddfdbbdef0ad0df8a2528381046233a2b6e21a/cpu/native/startup.c#L400

For some reason it works, if `zep_dispatch` get's there first, but if the `gnrc_border_router` get's to bind first, `zep_dispatch` may yield "EADDRINUSE".

Comment (benpicco) [reply dist/tools/zep_dispatch/start_network.sh:100]: Ah good catch! 

I removed the default local port, that doesn't work with multiple socket_zep instances anyway.

**Comment by miri64** [comment]
I would prefer to make the ZEP changes regarding local binding a separate PR. Same for the HELLO packet.

**Comment by benpicco** [comment]
Ok, I dropped them.
Should I change the default port? Or add a delay after the `zep_dispatcher`?

File: dist/tools/zep_dispatch/start_network.sh (hunk 12 above)
Comment (benpicco) [reply dist/tools/zep_dispatch/start_network.sh:100]: Should be solved now that the ZEP clients don't always listen by default :)

Comment (miri64) [reply dist/tools/zep_dispatch/start_network.sh:100]: Can you rebase then, please. You may also squash.

**Comment by miri64** [comment]
What happened to the HELLO packet now?

**Comment by benpicco** [comment]
> What happened to the HELLO packet now?

I added it here for convenience, can split it off again.

File: cpu/native/socket_zep/socket_zep.c [hunk 13]
```diff
@@ -390,6 +393,19 @@ static int _connect_remote(socket_zep_t *dev, const socket_zep_params_t *params)
...
+{
+    /* dummy packet */
+    zep_v2_data_hdr_t hdr = {
+        .hdr.preamble = "EX",
+        .hdr.version  = 2,
+        .type = SOCKET_ZEP_V2_TYPE_HELLO,
+        .resv = "HELLO",
+    };
+
+    real_write(dev->sock_fd, &hdr, sizeof(hdr));
```
Comment (miri64) [inline cpu/native/socket_zep/socket_zep.c:406]: Why use `real_write` and not `real_send` here?

**Comment by miri64** [comment]
> > What happened to the HELLO packet now?
> 
> I added it here for convenience, can split it off again.

Since this is to initiate contact with the dispatcher introduced here, I am fine with it to keep it here.

File: cpu/native/socket_zep/socket_zep.c (hunk 13 above)
Comment (benpicco) [reply cpu/native/socket_zep/socket_zep.c:406]: Why use `real_send` instead of `real_write`?

Comment (miri64) [reply cpu/native/socket_zep/socket_zep.c:406]: Because you are _sending_ a UDP packet...?

Comment (benpicco) [reply cpu/native/socket_zep/socket_zep.c:406]: But I'm *writing* to a socket

> With a zero flags argument, `send()` is equivalent to `write(2)`.

I can change it if you insist.

Comment (miri64) [reply cpu/native/socket_zep/socket_zep.c:406]: No, just wanted to know your reasoning.

File: dist/tools/zep_dispatch/start_network.sh [hunk 14]
```diff
@@ -0,0 +1,100 @@
+#!/bin/bash
```
Comment (bergzand) [inline dist/tools/zep_dispatch/start_network.sh:1]:
```suggestion
#!/usr/bin/env bash
```

File: dist/tools/zep_dispatch/start_network.sh (hunk 12 above)
Comment (bergzand) [inline dist/tools/zep_dispatch/start_network.sh:100]:
```suggestion
${NOSUDO} "${ELFFILE}" "${TAP}" "$@"
```

`"$@"` expands to `"$1" "$2" …` in bash

File: dist/tools/zep_dispatch/start_network.sh [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+else
+    USE_ZEP_DISPATCH=0
+fi
+
+ELFFILE=$1
+PREFIX=$2
+shift 2
+
+# tap will be the last argument
+for TAP in $@; do :; done
```
Comment (bergzand) [inline dist/tools/zep_dispatch/start_network.sh:78]: Either
```suggestion
for TAP in "$@"; do :; done
```
if you want to keep the script as portable as possible, or:
```suggestion
TAP="${@: -1}"
```
if you're using bashisms anyway.

File: dist/tools/zep_dispatch/start_network.sh [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+
+[ -z "${ELFFILE}" -o -z "${PREFIX}" -o -z "${TAP}" ] && {
```
Comment (bergzand) [inline dist/tools/zep_dispatch/start_network.sh:80]: Shellcheck suggests regular old `||` here, it claims that `-o` is not well defined.
```suggestion
[ -z "${ELFFILE}" || -z "${PREFIX}" || -z "${TAP}" ] && {
```

**Review by bergzand — COMMENTED** [review]
Shellcheck has some complaints here

**Comment by benpicco** [comment]
Can I squash?

**Comment by miri64** [comment]
What did you do, since I last look at this PR?

**Comment by miri64** [comment]
But I guess so. Seems like just cleanup of the script.

**Comment by benpicco** [comment]
I'm trying to reproduce the [failure of `tests/gnrc_netif_ieee802154`](https://ci.riot-os.org/RIOT-OS/RIOT/14755/0f9fe7b6ed1117375740c6f9eb480a1580363c91/output/compile/tests/gnrc_netif_ieee802154/native:gnu.txt) locally

    /tmp/dwq.0.305778657620171/0218f7b2bf1e71f3de1e9a4dd18186d8/build/tests_gnrc_netif_ieee802154.elf: open(/dev/net/tun): No such file or directory

However, on my machine `make all test` works on `native` with 

    sudo ip tuntap add tap0 mode tap user ${USER}
    sudo ip link set tap0 up

Is CI doing something different?
I don't understand how the changes result in `netdev_tap` being unable to open `/dev/net/tun`

**Comment by miri64** [comment]
There are no root permissions on the CI, so all tests using tap interfaces would not work. `tests/gnrc_netif_ieee802154` did not use any before this PR, so something in this PR changed that ;-).

**Comment by miri64** [comment]
However, if this PR that is supposed to _simplify_ the setup of an examples is getting so complicated, that it starts effecting some tests, it's a good sign, that this PR probably should be further split up...

**Comment by benpicco** [comment]
Not *that* complicated - I just fixed it *wrong*.
`tests/gnrc_netif_ieee802154` needs to disable the HELLO packet, it doesn't need `netdev_tap` (I falsely assumed that was used implicitly before)

File: cpu/native/socket_zep/socket_zep.c [hunk 17]
```diff
@@ -390,6 +393,23 @@ static int _connect_remote(socket_zep_t *dev, const socket_zep_params_t *params)
...
+#else
+    (void)dev;
+#endif
```
Comment (miri64) [inline cpu/native/socket_zep/socket_zep.c:410]:
```suggestion
    if (IS_USED(MODULE_SOCKET_ZEP_HELLO)) {
        /* dummy packet */
        zep_v2_data_hdr_t hdr = {
            .hdr.preamble = "EX",
            .hdr.version  = 2,
            .type = SOCKET_ZEP_V2_TYPE_HELLO,
            .resv = "HELLO",
        };

        real_write(dev->sock_fd, &hdr, sizeof(hdr));
    }
```

**Review by miri64 — CHANGES_REQUESTED** [review]
> tests/gnrc_netif_ieee802154 needs to disable the HELLO packet, […]

I guess you meant "needs not to enable the HELLO packet", because as far as I can see this (and hope it) is not a default module,

File: boards/native/Makefile.dep [hunk 18]
```diff
@@ -14,6 +16,7 @@ ifneq (,$(filter periph_can,$(FEATURES_USED)))
 endif
 
 ifneq (,$(filter socket_zep,$(USEMODULE)))
+  DEFAULT_MODULE += socket_zep_hello
```
Comment (miri64) [inline boards/native/Makefile.dep:19]: Oh... it is a default module... why?

File: boards/native/Makefile.dep [hunk 19]
```diff
@@ -1,5 +1,7 @@
 ifneq (,$(filter netdev_default,$(USEMODULE)))
-  USEMODULE += netdev_tap
+  ifeq (,$(filter socket_zep,$(USEMODULE)))
+    USEMODULE += netdev_tap
+  endif
```
Comment (miri64) [inline boards/native/Makefile.dep:4]: As far as I can see, this just makes things more complicated: Instead of having to include only `netdev_default` + `socket_zep`, we now need to include `netdev_default` (since we do that in the network applications for all boards) `netdev_tap` and `socket_zep`, for all applications that use networking with `native`. I prefer to keep this as is and thus not requiring no changes to the makefiles

Comment (benpicco) [reply boards/native/Makefile.dep:4]: But when I want to simulate a 802.15.4 node on native, I don't want it to also have a `netdev_tap` interface

File: boards/native/Makefile.dep (hunk 18 above)
Comment (benpicco) [reply boards/native/Makefile.dep:19]: There is no way to find out if an application should be used with a dispatcher. e.g. we could only add it to `examples/gnrc_border_router`, but then if other apps would use it, it would just mysteriously not work.

File: boards/native/Makefile.dep (hunk 19 above)
Comment (miri64) [reply boards/native/Makefile.dep:4]: But then you pick `socket_zep`. You have to pick `socket_zep` with this approach as well.

File: boards/native/Makefile.dep (hunk 18 above)
Comment (miri64) [reply boards/native/Makefile.dep:19]: Then this should be documented. It is a proprietary extension for our ZEP dispatcher. IMHO, this should be opt-in rather than opt-out.

File: boards/native/Makefile.dep (hunk 19 above)
Comment (benpicco) [reply boards/native/Makefile.dep:4]: So remove `netdev_default` and add `socket_zep` instead?
Yea I can do that.

Mind you that this behavior of 'overwriting' the default netdev is already used by [nrf52](https://github.com/RIOT-OS/RIOT/blob/master/boards/common/nrf52/nrf52840/Makefile.dep) and [esp*](https://github.com/RIOT-OS/RIOT/blob/master/cpu/esp_common/Makefile.dep#L37), I thought it would be straightforward to do the same for `native`.

File: boards/native/Makefile.dep (hunk 18 above)
Comment (benpicco) [reply boards/native/Makefile.dep:19]: I added a note to the documentation.

File: boards/native/Makefile.dep (hunk 19 above)
Comment (miri64) [reply boards/native/Makefile.dep:4]: > Mind you that this behavior of 'overwriting' the default netdev is already used by [nrf52](https://github.com/RIOT-OS/RIOT/blob/master/boards/common/nrf52/nrf52840/Makefile.dep) and [esp*](https://github.com/RIOT-OS/RIOT/blob/master/cpu/esp_common/Makefile.dep#L37), I thought it would be straightforward to do the same for `native`.

I know, but in this instance, for this specific PR, it seems making more fuzz then necessary. If you want, we can discuss this in a separate PR. I agree this should be more unified to other platforms, but as we already learned, this is affecting other applications than the one targeted by this PR, so we should rather discuss it in a different scope.

Comment (benpicco) [reply boards/native/Makefile.dep:4]: #15430

Comment (miri64) [reply boards/native/Makefile.dep:4]: Can this PR still be merged first?

Comment (miri64) [reply boards/native/Makefile.dep:4]: (with the changes of #15430 removed first of course)

File: boards/native/Makefile.include [hunk 20]
```diff
@@ -69,7 +69,7 @@ endif
 LINKFLAGS += -ffunction-sections
 
 # set the tap interface for term/valgrind
-ifneq (,$(filter netdev_default,$(USEMODULE)))
+ifneq (,$(filter netdev_tap,$(USEMODULE)))
```
Comment (miri64) [inline boards/native/Makefile.include:72]: IIRC this won't work BTW, since `netdev_tap` might not be selected at this point yet.

File: boards/native/Makefile.dep (hunk 18 above)
Comment (miri64) [reply boards/native/Makefile.dep:19]: Ermm... I meant that the opt-in mechanism should be documented in the README of `zep_dispatch` (and / or the `gnrc_border_router` example), not that the opt-out mechanism should be kept and stay in the documentation...

File: boards/native/Makefile.dep (hunk 19 above)
Comment (benpicco) [reply boards/native/Makefile.dep:4]: If you prefer it that way - I thought it would be easier to merge the smaller one first

Comment (miri64) [reply boards/native/Makefile.dep:4]: I think the smaller one is the bigger change, so I would like to get this one out of the way first and then discuss #15430.

File: cpu/native/include/socket_zep.h [hunk 21]
```diff
@@ -13,6 +13,16 @@
  *
  * @see @ref net_zep for protocol definitions
  *
+ * This ZEP implementation can send a dummy HELLO packet on startup.
+ * This is used to make dispatchers aware of the node.
+ * To enable this behavior, add
```
Comment (miri64) [inline cpu/native/include/socket_zep.h:18]: Hi, I am an implementer of a custom ZEP dispatcher. What are those HELLO dummy packets? How does the dispatcher recognize them as such? Does this belong to the ZEP standard? I don't find any information on those HELLO dummy packets. ;-)

Comment (benpicco) [reply cpu/native/include/socket_zep.h:18]: Well the dispatcher doesn't have to care -  my initial idea was for them to not have a ZEP header at all, so the dispatcher could just drop any non-ZEP traffic. If the dispatcher didn't care, it would be dropped latest by the ZEP implementation.

How if the dispatcher does parse the ZEP header to model more complex typologies, it also needs to know about the HELLO packet - I'll add a note.

Comment (miri64) [reply cpu/native/include/socket_zep.h:18]: Some specs on the ZEP header how it currently looks like would be nice either way ;-).

Comment (benpicco) [reply cpu/native/include/socket_zep.h:18]: Just for the HELLO or the ZEP header in general?

Comment (miri64) [reply cpu/native/include/socket_zep.h:18]: Just the HELLO packet should be enough.

Comment (benpicco) [reply cpu/native/include/socket_zep.h:18]: like so 6dd987ea7dd0298a020710c5c859a02a4a1deae1?

File: cpu/native/include/socket_zep.h [hunk 22]
```diff
@@ -26,6 +26,14 @@
  * A ZEP dispatcher can just drop those packets (ZEP type 0xFF) if it
  * chooses to parse the ZEP header.
  *
+ * The header of the HELLO packet will look like this:
+ *
+ * +-------------+---------+------+
+ * |    Magic    | Version | Type |
+ * +-------------+---------+------+
+ * | 0x45 | 0x58 |   0x2   | 0xFF |
+ * +-------------+---------+------+
```
Comment (miri64) [inline cpu/native/include/socket_zep.h:35]: Reserved field is missing.

```suggestion
 * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 * |         Preamble (EX)         |  Version (2)  |  Type  (255)  |
 * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 * |                                                               |
 * +                                                               +
 * |                                                               |
 * +                         Reserved (0)                         +
 * |                                                               |
 * +                                                               +
 * |                                                               |
 * +               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 * |               |      'H'      |      'E'      |      'L'      |
 * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 * |      'L'      |      'O'      |       0       |       0       |
 * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 * |       0       |       0       |       0       |       0       |
 * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

File: cpu/native/include/socket_zep.h [hunk 23]
```diff
@@ -13,6 +13,39 @@
...
+ * |                                                               |
+ * +                                                               +
+ * |                                                               |
+ * +               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * |               |      'H'      |      'E'      |      'L'      |
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * |      'L'      |      'O'      |       0       |       0       |
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * |       0       |       0       |       0       |       0       |
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```
Comment (miri64) [inline cpu/native/include/socket_zep.h:47]: Sorry, only now realized, that Doxygen won't like it this way. Other than that I am happy with the change and only would give it a final test ASAP. You may squash this PR.

```suggestion
 *     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 *     |         Preamble (EX)         |  Version (2)  |  Type  (255)  |
 *     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 *     |                                                               |
 *     +                                                               +
 *     |                                                               |
 *     +                         Reserved (0)                         +
 *     |                                                               |
 *     +                                                               +
 *     |                                                               |
 *     +               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 *     |               |      'H'      |      'E'      |      'L'      |
 *     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 *     |      'L'      |      'O'      |       0       |       0       |
 *     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 *     |       0       |       0       |       0       |       0       |
 *     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

Comment (benpicco) [reply cpu/native/include/socket_zep.h]: squashed.

**Comment by miri64** [comment]
Doc output is [looking good](https://5088-7370241-gh.circle-artifacts.com/0/doc/group__drivers__socket__zep.html)

**Review by miri64 — APPROVED** [review]
ACK, Testing procedures still work like a charm (for `gnrc_networking` I needed to set `TERMFLAGS="-z [::1]:17754"` though). Thanks for your patience during the review.

**Comment by benpicco** [comment]
Thank you for the review!
I had to comment out `USEMODULE += gnrc_netdev_default` for `examples/gnrc_networking` to test it

**Comment by miri64** [comment]
Oh yeah, me too. I replaced it with `USEMODULE += socket_zep`. But I guess this will be fixed with #15430.

**Comment by benpicco** [comment]
Ah that's why you had to add the TERMFLAGS - you can also do `USE_ZEP=1` and they will be added automatically.
